### PR TITLE
[Release notes] Add bundle-amend --remove option

### PR DESCRIPTION
**Pull request by lcawl**
## Summary

Add a `--remove` option to the `docs-builder changelog bundle-amend` command, so that changelogs can be removed from bundles post-release if necessary.

Previously, the bundle had to be manually edited or entirely replaced.

### Core changes

1. **Schema** — Added `exclude-entries` to [`Bundle.cs`](src/Elastic.Documentation/ReleaseNotes/Bundle.cs), [`BundleDto`](src/Elastic.Documentation.Configuration/ReleaseNotes/Bundle.cs), and [`ReleaseNotesSerialization.cs`](src/Elastic.Documentation.Configuration/ReleaseNotes/ReleaseNotesSerialization.cs).

2. **`BundleAmendMerger`** — New shared merger in [`BundleAmendMerger.cs`](src/Elastic.Documentation.Configuration/ReleaseNotes/BundleAmendMerger.cs): per amend file, apply exclusions then additions; ordered by amend number.

3. **Merge wiring** — Updated [`BundleLoader.cs`](src/Elastic.Documentation.Configuration/ReleaseNotes/BundleLoader.cs) and [`BundleValidationService.cs`](src/services/Elastic.Changelog/Rendering/BundleValidationService.cs) to use the merger.

4. **`ChangelogBundleAmendService`** — Supports `--remove`, `--force`, and `--dry-run`; writes `exclude-entries` to new `.amend-N.yaml` sidecars; checksum matching with optional `--force` for name-only removal.

5. **CLI** — [`ChangelogCommand.BundleAmend`](src/tooling/docs-builder/Commands/ChangelogCommand.cs) accepts `--remove`, `--force`, `--dry-run`; requires at least one of `--add` or `--remove`.

6. **`changelog remove`** — [`ChangelogRemoveService`](src/services/Elastic.Changelog/Bundling/ChangelogRemoveService.cs) uses the effective bundle (parent + amends − exclusions) for dependency checks; skips amend files as parent bundle candidates.

### Tests added

- [`BundleAmendMergerTests.cs`](tests/Elastic.Changelog.Tests/Changelogs/BundleAmendMergerTests.cs)
- Remove scenarios in [`BundleAmendTests.cs`](tests/Elastic.Changelog.Tests/Changelogs/BundleAmendTests.cs)
- Loader exclusion test in [`BundleLoaderTests.cs`](tests/Elastic.Changelog.Tests/Changelogs/BundleLoading/BundleLoaderTests.cs)
- Amend exclusion dependency test in [`ChangelogRemoveTests.cs`](tests/Elastic.Changelog.Tests/Changelogs/ChangelogRemoveTests.cs)
- Verify that `changelog render` omits excluded entry in `BundleValidationTests.cs` 
- Verify that `{changelog}` directive omits excluded entry and merged bundle has one entry in `ChangelogBasicTests.cs`

### Example usage

```sh
# Remove from effective bundle (parent stays untouched)
docs-builder changelog bundle-amend ./bundles/9.3.0.yaml \
  --remove ./docs/changelog/138723.yaml

# Replace in one amend
docs-builder changelog bundle-amend ./bundles/9.3.0.yaml \
  --remove ./docs/changelog/old.yaml \
  --add ./docs/changelog/new.yaml
```

## Steps for testing

I tested this command as follows:

1. Generate artifacts:
    ```sh
    ./build.sh clean
    ./build.sh publishbinaries
    ```
1. Check out a repo that has changelog bundles, for example, https://github.com/elastic/elastic-otel-java/tree/main
1. Run the `changelog bundle-amend` command to remove a changelog. For example, I tested against a bundle with a single changelog in the aforementioned repo:
    ```sh
    /path/docs-builder/.artifacts/publish/docs-builder/release/docs-builder changelog bundle-amend ./docs/releases/1.11.0.yaml --remove ./docs/changelog/upstream-update.yaml 
    ```
    The command failed with this message:
    ```sh
    Error: Bundle contains 'upstream-update.yaml' but with a different checksum than the file on disk. Re-create the bundle or use --force to remove by file name only.
    NOTE: /path/elastic-otel-java/docs/changelog/upstream-update.yaml
    ```
    When I re-ran the command with the `--force` option, it successfully created a `docs/releases/1.11.0.amend-1.yaml` file with the following contents:
    ```yaml
    exclude-entries:
    - file:
    name: upstream-update.yaml
    checksum: 3a0887aa3976c53f032f3952f33a95238fb55fa9
    ```
    The release note page that uses a changelog directive to render these bundles correctly shows the release (and its description) but the single changelog no longer appears:
    image

NOTE: If you try to remove a changelog from a bundle but you don't have access to the changelog file anymore, you get an error like this:

```sh
Error: File does not exist. Current directory: ... Tip: Repeat --remove for each file, or use comma-separated values (e.g., --remove "file1.yaml,file2.yaml"). Paths support tilde (~) expansion and can be relative or absolute.
NOTE: /path/GitHub/elastic-otel-java/835.yaml
```

## Generative AI disclosure

1. Did you use a generative AI (GenAI) tool to assist in creating this contribution?
- [x] Yes  
- [ ] No  
3. If you answered "Yes" to the previous question, please specify the tool(s) and model(s) used (e.g., Google Gemini, OpenAI ChatGPT-4, etc.).

Tool(s) and model(s) used: composer-2.5-fast

### CODE REVIEW

File: src/services/Elastic.Changelog/Bundling/ChangelogRemoveService.cs [hunk 1]
```diff
@@ -349,16 +350,26 @@ private async Task<IReadOnlyList<BundleDependency>> FindBundleDependenciesAsync(
...
 				var content = await _fileSystem.File.ReadAllTextAsync(bundleFile, ctx);
 				var bundle = ReleaseNotesSerialization.DeserializeBundle(content);
 
+				var amendPaths = ChangelogBundleAmendService.DiscoverAmendFiles(_fileSystem, bundleFile);
+				var amendBundles = new List<Bundle>();
+				foreach (var amendPath in amendPaths)
+				{
+					var amendContent = await _fileSystem.File.ReadAllTextAsync(amendPath, ctx);
+					amendBundles.Add(ReleaseNotesSerialization.DeserializeBundle(amendContent));
+				}
```
[the remark this answers is not in the document]
Comment (lcawl) [reply src/services/Elastic.Changelog/Bundling/ChangelogRemoveService.cs:375]: This was fixed in 63719884d27790b59a8ae417cfde65ef057f26b5 Per AI summary: "Per-amend deserialize wrapped in try/catch; corrupt amend files log a warning and dependency checks still use the parent bundle. Test: `Remove_WhenCorruptAmendBundle_StillBlocksFromParentReference`."

File: src/Elastic.Documentation.Configuration/ReleaseNotes/BundleLoader.cs [hunk 2]
```diff
@@ -262,63 +262,63 @@ private static LoadedBundle MergeBundleGroup(IGrouping<string, LoadedBundle> gro
...
+				parentBundle.Version,
+				parentBundle.Repo,
+				parentBundle.Owner,
+				mergedBundleData,
+				parentPath,
+				resolvedEntries);
+
+			foreach (var amend in group)
+				_ = mergedAmendPaths.Add(amend.FilePath);
 		}
```
[the remark this answers is not in the document]
Comment (lcawl) [reply src/Elastic.Documentation.Configuration/ReleaseNotes/BundleLoader.cs:375]: This BundleLoader orphan-amend change is declined. The local AI analysis is as follows:
"CodeRabbit suggests adding amend paths to mergedAmendPaths before continue when the parent bundle is missing, which removes orphan amend files from the output entirely.

That conflicts with intentional behavior:

- [cmd-bundle-amend.md](docs/cli/changelog/cmd-bundle-amend.md): “If an amend bundle is found without a matching parent bundle, it remains standalone.”
- Test [LoadBundles_AmendFileWithoutParent_RemainsStandalone](tests/Elastic.Changelog.Tests/Changelogs/BundleLoading/BundleLoaderTests.cs) expects a standalone bundle when only .amend-N.yaml exists."

The recommendation was to acknowledge the edge case (amend exists but parent failed to load) but keep standalone orphans; optionally add a follow-up to emit a warning without changing output shape.

File: src/Elastic.Documentation.Configuration/ReleaseNotes/BundleLoader.cs [hunk 3]
```diff
@@ -262,63 +262,63 @@ private static LoadedBundle MergeBundleGroup(IGrouping<string, LoadedBundle> gro
...
-			if (!mergedParents.TryGetValue(parentPath, out var mergedParent))
-				mergedParent = parentBundle;
+			var mergedEntryList = BundleAmendMerger.MergeEntries(parentBundle.Data.Entries, orderedAmendData);
+			var mergedBundleData = parentBundle.Data with { Entries = mergedEntryList };
 
-			// Merge the amend entries into the parent
-			var combinedEntries = mergedParent.Entries.Concat(amendBundle.Entries).ToList();
+			var bundleDirectory = fileSystem.Path.GetDirectoryName(parentPath) ?? string.Empty;
+			var changelogDirectory = fileSystem.Path.GetDirectoryName(bundleDirectory) ?? bundleDirectory;
+			var resolvedEntries = ResolveEntries(mergedBundleData, changelogDirectory, emitWarning);
```
Comment (Mpdreamz) [inline src/Elastic.Documentation.Configuration/ReleaseNotes/BundleLoader.cs:363]: The `changelogDirectory` fallback here is `bundleDirectory` (the bundles folder itself), but the per-bundle inference in `LoadBundles` at line 46–48 falls back to `bundlesFolderPath` when `GetDirectoryName` returns null — one level up from the bundle file, which is what `ResolveEntries` expects. These two paths are now inconsistent.

The cleaner fix is to pass `bundlesFolderPath` through to `MergeAmendFiles` and use it as the fallback:

```csharp
private List<LoadedBundle> MergeAmendFiles(
    List<LoadedBundle> bundles,
    string bundlesFolderPath,
    Action<string> emitWarning)
{
    ...
    var bundleDirectory = fileSystem.Path.GetDirectoryName(parentPath) ?? bundlesFolderPath;
    var changelogDirectory = fileSystem.Path.GetDirectoryName(bundleDirectory) ?? bundlesFolderPath;
    ...
}
```

Alternatively, store the resolved `changelogDirectory` on `LoadedBundle` when each bundle is first loaded, and reuse it during the merge — that avoids the double inference entirely.

File: src/services/Elastic.Changelog/Bundling/ChangelogBundleAmendService.cs [hunk 4]
```diff
@@ -170,101 +175,132 @@ public async Task<bool> AmendBundle(IDiagnosticsCollector collector, AmendBundle
...
 			};
 
 			var bundleForWrite = amendBundle;
-			if (linkAllowlistActive && shouldResolve)
+			if (entries.Count > 0 && shouldResolve && _configLoader != null)
 			{
-				ArgumentNullException.ThrowIfNull(parentBundleForAllowlist);
-				var owner = parentBundleForAllowlist.Products.Count > 0 ? parentBundleForAllowlist.Products[0].Owner ?? "elastic" : "elastic";
-				var repo = parentBundleForAllowlist.Products.Count > 0 ? parentBundleForAllowlist.Products[0].Repo : null;
+				var changelogConfig = await _configLoader.LoadChangelogConfiguration(collector, null, ctx);
```
Comment (Mpdreamz) [inline src/services/Elastic.Changelog/Bundling/ChangelogBundleAmendService.cs]: `LoadChangelogConfiguration` is called here a second time. It was already awaited at line 162 inside the `if (addFilePaths!.Count > 0)` block for the allowlist validation check. Hoist the config load and the derived `linkAllowRepos` / `linkAllowlistActive` values to a single call that covers both the validation and the sanitization pass.

**Review by Mpdreamz — CHANGES_REQUESTED** [review]
Good feature overall — the `BundleAmendMerger` extraction is clean and the test coverage is solid. Two issues worth fixing before merge.

File: src/services/Elastic.Changelog/Bundling/ChangelogBundleAmendService.cs (hunk 4 above)
Comment (lcawl) [reply src/services/Elastic.Changelog/Bundling/ChangelogBundleAmendService.cs]: Addressed in 1cd7a7db328aef969e123f3a0b585efb1a460f64:

- `LoadChangelogConfiguration` is loaded once and reused for allowlist validation and sanitization.

File: src/Elastic.Documentation.Configuration/ReleaseNotes/BundleLoader.cs (hunk 3 above)
Comment (lcawl) [reply src/Elastic.Documentation.Configuration/ReleaseNotes/BundleLoader.cs:363]: Addressed in 1cd7a7db328aef969e123f3a0b585efb1a460f64: 

- `MergeAmendFiles` takes `bundlesFolderPath` and uses the same fallback as `LoadBundles`; `LoadBundlesFromContent` passes `string.Empty` and `emitWarning`.

**Review by Mpdreamz — APPROVED** [review]
